### PR TITLE
don't allow MLB assigns with different axes

### PR DESCRIPTION
**Pull request by chaosagent**
this allows initializing axis-sharded weights without sharding on axis, and letting the shard happen implicitly after in the first step.

needed for #3472

### CODE REVIEW

**Comment by chenyuxyz** [comment]
i don't quite understand the pr title and the intended change? isn't the code change more limiting? also i think the test already passes without code change?

**Comment by chaosagent** [comment]
it doesn't pass, because the old buffers before axis shard are larger than the new buffers. Then .nbytes is not the same as prod(shape), and copyin will fail.

**Comment by chaosagent** [comment]
ok the test is fixed, this bug only happens if the axis=None buffer already had a realized buffer (eg when resuming training from ckpt)

**Comment by chenyuxyz** [comment]
Is there a way to make it more explicit? What would be the alternative if we asserts the axis to be the same.

While we don't have a full spec for assign now (#3395), it feels less predictable that an assignment changes shard axis implicitly. I would expect that the assignment eventually checks certain properties (like shape, size, shard axis, although there's more complicated issues like real size can be different or contiguous etc) and throws error accordingly.

File: test/test_multitensor.py [hunk 1]
```diff
@@ -307,6 +307,24 @@ def test_bn_ast_on_devices(self):
     # for i, ast in enumerate(asts):
     #   print(f"{i} {ast}")
 
+  def test_mlb_assign_change_axis(self):
+    devices = (d0, d1, d2, d3)
+
+    # axis=None <- axis=None + axis=0
+    t_none = Tensor.zeros((16, 16)).shard(devices).realize()  # allocate 16*16 length buffers on each device
+    assert t_none.lazydata.lbs[0].realized.size == 16 * 16  # check that buffers have been allocated
```
Comment (chenyuxyz) [inline test/test_multitensor.py:315]: const multi Tensor will not have realized (similar to single LB case) so this might change

**Review by chenyuxyz — COMMENTED** [review]
Let me think more about the desired case. Can you fix the conflict in the meantime?

**Comment by chenyuxyz** [comment]
> But when it get to that point, the optimizer has already gotten the reference to the Tensor, and thus we need a way to swap out the lazydata that isn't assign

We can shard the model weights before putting it in optimizer. Does that work?

**Comment by chenyuxyz** [comment]
this fails with the newly added unsynced bn with virtual multi devices like llvm and clang

**Pull request merged**